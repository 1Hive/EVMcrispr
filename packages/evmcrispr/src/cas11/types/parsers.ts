import type { Parser } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

import type { Node } from './ast';

export type NodeParser<T = Node> = Parser<T, string, NodeParserState>;
