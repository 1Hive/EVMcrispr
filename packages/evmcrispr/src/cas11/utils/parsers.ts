import type { Err } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

export const buildParserError = (
  { index, error }: Err<string, null> | Err<string, NodeParserState>,
  type: string,
  msg?: string,
): string => {
  const splitRes = error.split('got');
  const wrongValueEncountered =
    splitRes.length === 2 ? splitRes[1].trim() : null;

  const errorMessage = msg
    ? `${msg}${wrongValueEncountered ? `, got ${wrongValueEncountered}` : ''}`
    : error.split('): ')[1];
  return `${type}(col: ${index}): ${errorMessage}`;
};

export const getIncorrectReceivedValue = (errorMsg: string): string => {
  const splitRes = errorMsg.split('got ');

  if (splitRes.length === 2) {
    return `, got ${splitRes[1].trim()}`;
  }

  return '';
};
