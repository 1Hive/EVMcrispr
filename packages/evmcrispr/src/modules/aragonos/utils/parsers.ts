import type { PublicClient } from "viem";
import { parseAbiItem } from "viem";

import type { Address } from "../../../types";
import type { ParsedApp, Repo } from "../types";
import { getSystemApp, isSystemApp } from "./interfaces";

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return "";
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : "";
};

const fetchImplementationAddress = (
  appAddress: Address,
  client: PublicClient,
): Promise<Address> => {
  return client.readContract({
    address: appAddress,
    abi: [
      parseAbiItem("function implementation() public view returns (address)"),
    ],
    functionName: "implementation",
  });
};

export const parseApp = async (
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  app: any,
  client: PublicClient,
): Promise<ParsedApp> => {
  const { address, appId, roles, version } = app;
  const { name: repoName, registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress:
      codeAddress ?? (await fetchImplementationAddress(address, client)),
    contentUri,
    name,
    registryName: registry?.name,
    roles: (roles as any[]).map((role) => ({
      ...role,
      roleHash: role["hash"] ?? role["roleHash"],
    })),
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};
