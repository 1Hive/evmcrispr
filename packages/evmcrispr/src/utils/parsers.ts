import { getSystemApp, isSystemApp } from '.';
import type { ParsedApp, Repo } from '../types';

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return '';
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : '';
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseApp = (app: any): ParsedApp => {
  const { address, appId, implementation, repoName, roles, version } = app;
  const { registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress: codeAddress ?? implementation.address,
    contentUri,
    name,
    registryName: registry?.name,
    roles,
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};

export const timeUnits: { [key: string]: number } = {
  s: 1,
  m: 60,
  h: 3600,
  d: 86400,
  w: 604800,
  mo: 2592000,
  y: 31536000,
};
