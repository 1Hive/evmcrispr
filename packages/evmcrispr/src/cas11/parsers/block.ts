import type { Parser } from 'arcsecond';
import { char, coroutine, recursiveParser } from 'arcsecond';

import type { BlockExpressionNode } from '../types';
import { NodeType } from '../types';
import { commandExpressionParser } from './command';
import {
  endOfLine,
  optionalEmptyLines,
  optionalWhitespace,
  surroundedBy,
} from './utils';

export const blockExpressionParser: Parser<BlockExpressionNode, string, any> =
  recursiveParser(() =>
    coroutine(function* () {
      yield char('(');

      yield optionalWhitespace;

      yield endOfLine;

      const scopedCommands = yield optionalEmptyLines(commandExpressionParser);

      yield surroundedBy(optionalWhitespace)(char(')'));

      return {
        type: NodeType.BlockExpression,
        body: scopedCommands,
      } as unknown as BlockExpressionNode;
    }),
  );
