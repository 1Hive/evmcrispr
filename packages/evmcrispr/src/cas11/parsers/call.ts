import type { Parser } from 'arcsecond';
import {
  char,
  choice,
  coroutine,
  letters,
  possibly,
  recursiveParser,
} from 'arcsecond';

import type { StringLiteralNode } from '../types';
import { NodeType } from '../types';

import {
  addressParser,
  probableIdentifierParser,
  variableIdentifierParser,
} from './primaries';
import { argumentsParser } from './expression';
import { helperFunctionParser } from './helper';

export const callExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    coroutine(function* () {
      const target = yield choice([
        addressParser,
        variableIdentifierParser,
        helperFunctionParser,
        probableIdentifierParser,
      ]);

      let res = yield char(':');
      let callExpressionNode: any = target;

      do {
        const callee = yield letters.map(
          (value): StringLiteralNode => ({
            type: NodeType.StringLiteral,
            value,
          }),
        );

        const args = yield argumentsParser;

        callExpressionNode = {
          type: NodeType.CallExpression,
          target: callExpressionNode,
          callee,
          args,
        };

        res = yield possibly(char(':'));
      } while (res);

      return callExpressionNode;
    }),
);
