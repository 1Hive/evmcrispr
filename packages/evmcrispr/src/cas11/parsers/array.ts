import type { Parser } from 'arcsecond';
import {
  between,
  char,
  coroutine,
  recursiveParser,
  sequenceOf,
} from 'arcsecond';

import type { ArrayExpressionNode } from '../types';
import { NodeType } from '../types';

import { argumentExpressionParser } from './expression';

import { commaSeparated, optionalWhitespace } from './utils';

export const arrayExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    between(sequenceOf([char('['), optionalWhitespace]))(
      sequenceOf([optionalWhitespace, char(']')]),
    )(commaSeparated(argumentExpressionParser)).map(
      (elements: any): ArrayExpressionNode => {
        return {
          type: NodeType.ArrayExpression,
          elements,
        };
      },
    ),
);

export const arrayExpresionParser = recursiveParser(() =>
  coroutine(function* () {
    yield char('[');
    yield optionalWhitespace;
    const elements = yield commaSeparated(argumentExpressionParser);
    yield optionalWhitespace;
    yield char(']');

    return {
      type: NodeType.ArrayExpression,
      elements,
    };
  }),
);
