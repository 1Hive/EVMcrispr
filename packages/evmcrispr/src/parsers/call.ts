import type { Parser } from 'arcsecond';
import {
  choice,
  coroutine,
  letters,
  possibly,
  recursiveParser,
  regex,
} from 'arcsecond';

import type {
  CallExpressionNode,
  LocationData,
  NodeParser,
  NodeParserState,
} from '../types';
import { NodeType } from '../types';

import {
  addressParser,
  probableIdentifierParser,
  variableIdentifierParser,
} from './primaries';
import { argumentsParser } from './expression';
import { helperFunctionParser } from './helper';
import {
  callOperatorParser,
  createNodeLocation,
  currentContexDataParser,
} from './utils';
import { buildParserError } from '../utils/parsers';

const chainedCallExpressionParser = (
  target: CallExpressionNode,
): Parser<CallExpressionNode, string, NodeParserState> =>
  recursiveParser(() =>
    coroutine(run => {
      const initialContext: LocationData =
        run(currentContexDataParser);

      const method: CallExpressionNode['method'] = run(letters);

      const args: CallExpressionNode['args'] =
        run(argumentsParser);

      const finalContext: LocationData =
        run(currentContexDataParser);

      const n: CallExpressionNode = {
        type: NodeType.CallExpression,
        target,
        method,
        args,
        loc: createNodeLocation(initialContext, finalContext),
      };

      // Check for further chained call expressions
      if (run(possibly(callOperatorParser))) {
        const chainedNode: CallExpressionNode = run(chainedCallExpressionParser(
          n,
        ));
        return chainedNode;
      }

      return n;
    }),
  );

const enclosingParsers = [callOperatorParser];

const callableExpressions = choice([
  addressParser(enclosingParsers),
  variableIdentifierParser(enclosingParsers),
  helperFunctionParser,
  probableIdentifierParser(enclosingParsers),
]);

export const callExpressionParser: NodeParser<CallExpressionNode> =
  recursiveParser(() =>
    coroutine(run => {
      const initialContext: LocationData =
        run(currentContexDataParser);
      const target: CallExpressionNode['target'] =
        run(callableExpressions);

      run(callOperatorParser);

      const methodRegex = /^[a-zA-Z_{1}][a-zA-Z0-9_]+/;
      const method: CallExpressionNode['method'] = run(regex(
        methodRegex,
      ));
      const args: CallExpressionNode['args'] = run(argumentsParser.errorMap((err) =>
        buildParserError(err, ''),
      ));

      const finalContext: LocationData =
        run(currentContexDataParser);

      const n: CallExpressionNode = {
        type: NodeType.CallExpression,
        target,
        method,
        args,
        loc: createNodeLocation(initialContext, finalContext),
      };

      // Check for chained call expressions
      if (run(possibly(callOperatorParser))) {
        const chainedCallNode: CallExpressionNode = run(chainedCallExpressionParser(
          n,
        ));

        return chainedCallNode;
      }

      return n;
    }),
  );
