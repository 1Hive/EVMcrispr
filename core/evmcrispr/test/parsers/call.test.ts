import type { Case } from '@1hive/evmcrispr-test-common';
import { runParser } from '@1hive/evmcrispr-test-common';

import { callExpressionParser } from '../../src/parsers/call';

describe('Parsers - call expression', () => {
  describe.concurrent.each<Case>([
    {
      title: 'made from a literal address',
      value: `0x14FA5C16Af56190239B997485656F5c8b4f86c4b::getEntry(0, @token(WETH))`,
    },
    {
      title: 'containing nested call expression',
      value: `$superfluid::createFlow(@token("DAIx"), $finance::vault([1,2,3]), $contract::method(), 10e18m, 'this is a nice description')`,
    },
    {
      title: 'made from a helper function',
      value: `@token(DAIx)::upgrade(@token(DAI), 1800e18)`,
    },
    {
      title: 'with following chained call expressions',
      value: `$registryContract::getToken(1)::approve(@me, 560.25e18)::another()`,
    },
  ])('', ({ title, value }) => {
    it(`should parse a call expression ${title} correctly`, () => {
      const parsedValue = runParser(callExpressionParser, value);

      expect(parsedValue).toMatchSnapshot();
    });
  });
});
