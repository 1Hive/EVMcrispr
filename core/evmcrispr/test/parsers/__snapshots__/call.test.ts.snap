// Vitest Snapshot v1

exports[`Parsers - call expression > should parse a call expression containing nested call expression correctly 1`] = `
{
  "args": [
    {
      "args": [
        {
          "loc": {
            "end": {
              "col": 37,
              "line": 1,
            },
            "start": {
              "col": 31,
              "line": 1,
            },
          },
          "type": "StringLiteral",
          "value": "DAIx",
        },
      ],
      "loc": {
        "end": {
          "col": 38,
          "line": 1,
        },
        "start": {
          "col": 24,
          "line": 1,
        },
      },
      "name": "token",
      "type": "HelperFunctionExpression",
    },
    {
      "args": [
        {
          "elements": [
            {
              "loc": {
                "end": {
                  "col": 58,
                  "line": 1,
                },
                "start": {
                  "col": 57,
                  "line": 1,
                },
              },
              "type": "NumberLiteral",
              "value": "1",
            },
            {
              "loc": {
                "end": {
                  "col": 60,
                  "line": 1,
                },
                "start": {
                  "col": 59,
                  "line": 1,
                },
              },
              "type": "NumberLiteral",
              "value": "2",
            },
            {
              "loc": {
                "end": {
                  "col": 62,
                  "line": 1,
                },
                "start": {
                  "col": 61,
                  "line": 1,
                },
              },
              "type": "NumberLiteral",
              "value": "3",
            },
          ],
          "loc": {
            "end": {
              "col": 63,
              "line": 1,
            },
            "start": {
              "col": 56,
              "line": 1,
            },
          },
          "type": "ArrayExpression",
        },
      ],
      "loc": {
        "end": {
          "col": 64,
          "line": 1,
        },
        "start": {
          "col": 40,
          "line": 1,
        },
      },
      "method": "vault",
      "target": {
        "loc": {
          "end": {
            "col": 48,
            "line": 1,
          },
          "start": {
            "col": 40,
            "line": 1,
          },
        },
        "type": "VariableIdentifier",
        "value": "$finance",
      },
      "type": "CallExpression",
    },
    {
      "args": [],
      "loc": {
        "end": {
          "col": 85,
          "line": 1,
        },
        "start": {
          "col": 66,
          "line": 1,
        },
      },
      "method": "method",
      "target": {
        "loc": {
          "end": {
            "col": 75,
            "line": 1,
          },
          "start": {
            "col": 66,
            "line": 1,
          },
        },
        "type": "VariableIdentifier",
        "value": "$contract",
      },
      "type": "CallExpression",
    },
    {
      "loc": {
        "end": {
          "col": 93,
          "line": 1,
        },
        "start": {
          "col": 87,
          "line": 1,
        },
      },
      "power": 18,
      "timeUnit": "m",
      "type": "NumberLiteral",
      "value": "10",
    },
    {
      "loc": {
        "end": {
          "col": 123,
          "line": 1,
        },
        "start": {
          "col": 95,
          "line": 1,
        },
      },
      "type": "StringLiteral",
      "value": "this is a nice description",
    },
  ],
  "loc": {
    "end": {
      "col": 124,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "method": "createFlow",
  "target": {
    "loc": {
      "end": {
        "col": 11,
        "line": 1,
      },
      "start": {
        "col": 0,
        "line": 1,
      },
    },
    "type": "VariableIdentifier",
    "value": "$superfluid",
  },
  "type": "CallExpression",
}
`;

exports[`Parsers - call expression > should parse a call expression made from a helper function correctly 1`] = `
{
  "args": [
    {
      "args": [
        {
          "loc": {
            "end": {
              "col": 32,
              "line": 1,
            },
            "start": {
              "col": 29,
              "line": 1,
            },
          },
          "type": "ProbableIdentifier",
          "value": "DAI",
        },
      ],
      "loc": {
        "end": {
          "col": 33,
          "line": 1,
        },
        "start": {
          "col": 22,
          "line": 1,
        },
      },
      "name": "token",
      "type": "HelperFunctionExpression",
    },
    {
      "loc": {
        "end": {
          "col": 42,
          "line": 1,
        },
        "start": {
          "col": 35,
          "line": 1,
        },
      },
      "power": 18,
      "type": "NumberLiteral",
      "value": "1800",
    },
  ],
  "loc": {
    "end": {
      "col": 43,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "method": "upgrade",
  "target": {
    "args": [
      {
        "loc": {
          "end": {
            "col": 11,
            "line": 1,
          },
          "start": {
            "col": 7,
            "line": 1,
          },
        },
        "type": "ProbableIdentifier",
        "value": "DAIx",
      },
    ],
    "loc": {
      "end": {
        "col": 12,
        "line": 1,
      },
      "start": {
        "col": 0,
        "line": 1,
      },
    },
    "name": "token",
    "type": "HelperFunctionExpression",
  },
  "type": "CallExpression",
}
`;

exports[`Parsers - call expression > should parse a call expression made from a literal address correctly 1`] = `
{
  "args": [
    {
      "loc": {
        "end": {
          "col": 54,
          "line": 1,
        },
        "start": {
          "col": 53,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "0",
    },
    {
      "args": [
        {
          "loc": {
            "end": {
              "col": 67,
              "line": 1,
            },
            "start": {
              "col": 63,
              "line": 1,
            },
          },
          "type": "ProbableIdentifier",
          "value": "WETH",
        },
      ],
      "loc": {
        "end": {
          "col": 68,
          "line": 1,
        },
        "start": {
          "col": 56,
          "line": 1,
        },
      },
      "name": "token",
      "type": "HelperFunctionExpression",
    },
  ],
  "loc": {
    "end": {
      "col": 69,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "method": "getEntry",
  "target": {
    "loc": {
      "end": {
        "col": 42,
        "line": 1,
      },
      "start": {
        "col": 0,
        "line": 1,
      },
    },
    "type": "AddressLiteral",
    "value": "0x14FA5C16Af56190239B997485656F5c8b4f86c4b",
  },
  "type": "CallExpression",
}
`;

exports[`Parsers - call expression > should parse a call expression with following chained call expressions correctly 1`] = `
{
  "args": [],
  "loc": {
    "end": {
      "col": 66,
      "line": 1,
    },
    "start": {
      "col": 57,
      "line": 1,
    },
  },
  "method": "another",
  "target": {
    "args": [
      {
        "args": [],
        "loc": {
          "end": {
            "col": 43,
            "line": 1,
          },
          "start": {
            "col": 40,
            "line": 1,
          },
        },
        "name": "me",
        "type": "HelperFunctionExpression",
      },
      {
        "loc": {
          "end": {
            "col": 54,
            "line": 1,
          },
          "start": {
            "col": 45,
            "line": 1,
          },
        },
        "power": 18,
        "type": "NumberLiteral",
        "value": "560.25",
      },
    ],
    "loc": {
      "end": {
        "col": 55,
        "line": 1,
      },
      "start": {
        "col": 32,
        "line": 1,
      },
    },
    "method": "approve",
    "target": {
      "args": [
        {
          "loc": {
            "end": {
              "col": 29,
              "line": 1,
            },
            "start": {
              "col": 28,
              "line": 1,
            },
          },
          "type": "NumberLiteral",
          "value": "1",
        },
      ],
      "loc": {
        "end": {
          "col": 30,
          "line": 1,
        },
        "start": {
          "col": 0,
          "line": 1,
        },
      },
      "method": "getToken",
      "target": {
        "loc": {
          "end": {
            "col": 17,
            "line": 1,
          },
          "start": {
            "col": 0,
            "line": 1,
          },
        },
        "type": "VariableIdentifier",
        "value": "$registryContract",
      },
      "type": "CallExpression",
    },
    "type": "CallExpression",
  },
  "type": "CallExpression",
}
`;
